specs/experimental/op-contracts-manager.md

```diff
@@ -153,27 +166,119 @@ This provides the following benefits:
+
+1. The Upgrade Controller Safe will `DELEGATECALL` to the `OPCM.upgrade()` method.
+2. For each `_systemConfig`, the list of addresses in the chain is retrieved.
+3. For each address, a two step upgrade is used where:
+   1. the first upgrade is to an `InitializerResetter` which resets the `initialized` value.
+   1. the implementation is updated to the final address and `upgrade()` is called on that address.
+
+This approach requires that all contracts have an `upgrade()` function which sets the `initialized`
+value to `true`. The `upgrade` function body should be empty unless it is used to set a new state
+variable added to that contract since the last upgrade.
+
+#### `NewChainConfig` struct
+
+This struct is used to pass the new chain configuration to the `upgrade` method, and so it will
+vary for each release of the OP Contracts Manager, based on what (if any) new parameters are added.
+
+In practice, this struct is likely to be have a unique name for each release of the OP Contracts
+Manager.
+
+#### Requirements on the OP Chain contracts
+
+In general, all contracts used in an OP Chain SHOULD be proxied with a single shared implementation.
+This means that all values which are not constant across OP Chains SHOULD be held in storage rather
+than the bytecode of the implementation.
+
+Any contracts which do not meet this requirement will need to be deployed by the `upgrade()`
+function, increasing the cost and reducing the number of OP Chains which can be atomically upgraded.
+
+## Adding game types
+
+Because different OP Chains within a Superchain may use different dispute game types, and are
+expected to move from a permissioned to permissionless game over time, an `addGameType` method is
+provided to enable adding a new game type to multiple games at once.
+
+### Interface
+
+#### `addGameType`
+
+The `addGameType` method is used to orchestrate the actions required to add a new game type to one
+or more chains.
+
+```solidity
+struct NewGameConfig {
+  // fields will vary depending on the game type
+}
+
+function addGameType(ISystemConfig[] _systemConfigs, NewGameConfig[] _newGames) public;
+```
+
+### Implementation
+
+The high level logic of the `addGameType` method is as follows (for each chain):
+
+1. Deploys the new game Creator contract for that game type.
+2. Calls `setImplementation()` on the `DisputeGameFactory`
+3. Calls `setAnchorState()` on the `AnchorStateRegistry`
+
+Note that in [Standard Configuration chains](../protocol/superchain-configuration.md):
+
+- `DisputeGameFactory.setImplementation()` is authorized to the Upgrade Controller
+- `AnchorStateRegistry.setAnchorState()` is authorized to the Guardian
+
+Since the Guardian is its own Safe controlled by the Security Council, and the Upgrade Controller is
+a 2 of 2 jointly controlled by the Security Council and Optimism Foundation, this is not impossible
+to implement, but will be ugly and require new tooling and processes to do correctly.
```

Comment:
As mentioned in the description, I'm doubting the value of this method vs. just having reusable superchain-ops tasks for adding a new game. I'm generally in favor of scrapping addGameType() for now.

Reply:
Of course we could add two separate functions which basically do the different steps if it's deemed valuable, it's just not clear to me that it does much to reduce the operational lift.

Reply:
The key requirement is having some way to actually deploy the `FaultDisputeGame` contract.  The superchain-ops task is easy enough but it only calls `setImplementation` and requires a deployed `FaultDisputeGame` as input. superchain-ops doesn't handle that because you need to be sure to deploy the right version of the contract and OPCM is designed to be the on-chain source of "the right version" (as I understand it).

An `op-deployer` bootstrap command could be used to do the deployment but we need to be crystal clear on what the process is and get it tested so that it works with every contracts release.  The bootstrap scripts for op-deployer have to be in the contracts release and working so it's critical they're actually tested.

Reply:
Also for the record, when adding a new game type you currently need to reinitialise the `AnchorStateRegistry` - `setAnchorState` can only set the anchor state to a game of the same type that resolved as valid (allowing you to rollback a bad update).  The superchain-ops template can do that re-initialization already and there's work in place to remove the requirement by having a single anchor state used for all game types (that's updated after the air-gap so it's dependable). We aren't blocking on that for Isthmus (it may be ready, not sure) but happy to leave that to the superchain-ops task.

Reply:
> The key requirement is having some way to actually deploy the `FaultDisputeGame` contract.

Ah, I see. I think that along with https://github.com/ethereum-optimism/design-docs/pull/153, that it should be quite doable to deploy and initialize a new proxy to a `Creator` contract implementation.

> when adding a new game type you currently need to reinitialise the `AnchorStateRegistry`

Got it. If reinitializing is the right approach, then we can also do that in the same function on the basis that it would be called by the Upgrade Controller, and afaict that overcomes all my concerns, thanks for clarifying!

Reply:
Updated the spec in this commit: 34e18e4de0d2